test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -811,13 +814,59 @@ func runWorkload(ctx context.Context, b *testing.B, tc *testCase, w *workload, c
 	numPodsScheduledPerNamespace := make(map[string]int)
 
 	if cleanup {
-		b.Cleanup(func() {
+		// This must run before controllers get cleaned up.
+		defer func() {
+			deleteNow := *metav1.NewDeleteOptions(0)
 			for namespace := range numPodsScheduledPerNamespace {
-				if err := client.CoreV1().Namespaces().Delete(ctx, namespace, metav1.DeleteOptions{}); err != nil {
-					b.Errorf("Deleting Namespace in numPodsScheduledPerNamespace: %v", err)
+				// Pods have to be deleted explicitly, with no grace period. Normally
+				// kubelet will set the DeletionGracePeriodSeconds to zero when it's okay
+				// to remove a deleted pod, but we don't run kubelet...
+				pods, err := client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+				if err != nil {
+					b.Fatalf("failed to list pods in %q: %v", namespace, err)
+				}
+				for _, pod := range pods.Items {
+					if err := client.CoreV1().Pods(namespace).Delete(ctx, pod.Name, deleteNow); err != nil {
+						b.Fatalf("failed to delete pod %q in namespace %q: %v", pod.Name, namespace, err)
+					}
+				}
+				if err := client.CoreV1().Namespaces().Delete(ctx, namespace, deleteNow); err != nil {
+					b.Fatalf("Deleting Namespace in numPodsScheduledPerNamespace: %v", err)
 				}
 			}
-		})
+			gomega.NewGomegaWithT(b).Eventually(ctx, func(ctx context.Context) ([]interface{}, error) {
+				var objects []interface{}
+				namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
```

Comment:
If `cleanup` is true, doesn't that mean **`namespaces` is empty** is enough? Or at least we can directly return when we want to append below, because that means the objects isn't empty.

Reply:
The purpose here is to provide a complete report of everything that hasn't been removed when removal doesn't finish in time. That is more accurate than reporting just one object, which might give the wrong impression.

Not sure whether that addresses your concern?

Reply:
Benefit to debugging then.